File: examples/upload_examples.js
#!/usr/bin/env node

/**
 * AdapterOS Upload Examples
 *
 * Complete working examples for uploading .aos adapters to AdapterOS.
 * Run with: node upload_examples.js
 *
 * Requirements:
 *   npm install form-data node-fetch
 */

const fs = require('fs');
const path = require('path');
const FormData = require('form-data');
const fetch = require('node-fetch');

/**
 * Simple uploader - minimal functionality
 */
class SimpleUploader {
    constructor(baseUrl, token) {
        this.baseUrl = baseUrl;
        this.token = token;
    }

    async upload(filePath, name) {
        const form = new FormData();
        form.append('file', fs.createReadStream(filePath), {
            filename: path.basename(filePath),
        });
        form.append('name', name);

        const response = await fetch(
            `${this.baseUrl}/v1/adapters/upload-aos`,
            {
                method: 'POST',
                headers: {
                    'Authorization': `Bearer ${this.token}`,
                    ...form.getHeaders(),
                },
                body: form,
            }
        );

        if (!response.ok) {
            const error = await response.json();
            throw new Error(
                `Upload failed (${response.status}): ` +
                `${error.error_code} - ${error.message}`
            );
        }

        return response.json();
    }
}

/**
 * Production uploader - with retry, validation, and progress
 */
class ProductionUploader {
    constructor(baseUrl, token, options = {}) {
        this.baseUrl = baseUrl;
        this.token = token;
        this.maxRetries = options.maxRetries || 3;
        this.timeout = options.timeout || 60000; // 60 seconds
        this.retryDelay = options.retryDelay || 1000; // 1 second
    }

    /**
     * Validate file before upload
     */
    validateFile(filePath) {
        // Check existence
        if (!fs.existsSync(filePath)) {
            throw new Error(`File not found: ${filePath}`);
        }

        // Check extension
        if (!filePath.endsWith('.aos')) {
            throw new Error(`File must have .aos extension, got: ${filePath}`);
        }

        // Check size
        const stats = fs.statSync(filePath);
        const maxSize = 1024 * 1024 * 1024; // 1GB
        if (stats.size > maxSize) {
            throw new Error(
                `File too large: ${(stats.size / (1024 * 1024)).toFixed(1)}MB ` +
                `(max: ${maxSize / (1024 * 1024)}MB)`
            );
        }

        // Validate .aos structure
        this.validateAosStructure(filePath);
    }

    /**
     * Validate .aos file format
     */
    validateAosStructure(filePath) {
        const buffer = Buffer.alloc(8);
        const fd = fs.openSync(filePath, 'r');

        try {
            // Read header
            fs.readSync(fd, buffer, 0, 8, 0);

            // Parse header (little-endian u32 values)
            const offset = buffer.readUInt32LE(0);
            const length = buffer.readUInt32LE(4);

            // Validate header
            if (offset < 8 || length === 0) {
                throw new Error(
                    `Invalid .aos header: offset=${offset}, length=${length}`
                );
            }

            // Check manifest fits in file
            const stats = fs.fstatSync(fd);
            if (offset + length > stats.size) {
                throw new Error(
                    `Manifest extends beyond file: ` +
                    `offset=${offset}, length=${length}, size=${stats.size}`
                );
            }

            // Validate manifest is JSON
            const manifestBuffer = Buffer.alloc(length);
            fs.readSync(fd, manifestBuffer, 0, length, offset);
            const manifest = JSON.parse(manifestBuffer.toString('utf8'));

            if (typeof manifest !== 'object' || Array.isArray(manifest)) {
                throw new Error('Manifest must be JSON object');
            }
        } finally {
            fs.closeSync(fd);
        }
    }

    /**
     * Upload adapter with validation and retry
     */
    async upload(filePath, name, options = {}) {
        // Validate input
        this.validateFile(filePath);

        const {
            description,
            tier = 'ephemeral',
            category = 'general',
            scope = 'general',
            rank = 1,
            alpha = 1.0,
            onProgress,
        } = options;

        // Validate parameters
        if (rank < 1 || rank > 512) {
            throw new Error(`Rank must be 1-512, got ${rank}`);
        }
        if (alpha < 0.0 || alpha > 100.0) {
            throw new Error(`Alpha must be 0.0-100.0, got ${alpha}`);
        }

        // Retry loop
        let lastError;
        for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
            try {
                return await this.doUpload(
                    filePath,
                    name,
                    description,
                    tier,
                    category,
                    scope,
                    rank,
                    alpha,
                    onProgress
                );
            } catch (error) {
                lastError = error;

                if (attempt < this.maxRetries) {
                    const isRetryable =
                        error.code === 'ETIMEDOUT' ||
                        error.code === 'ECONNREFUSED' ||
                        error.code === 'ENOTFOUND';

                    if (isRetryable) {
                        const delay = Math.pow(2, attempt) * this.retryDelay;
                        console.log(
                            `Attempt ${attempt} failed, retrying in ${delay}ms...`
                        );
                        await this.sleep(delay);
                        continue;
                    }
                }

                throw error;
            }
        }

        throw new Error(`Upload failed after ${this.maxRetries} attempts: ${lastError}`);
    }

    /**
     * Perform actual upload
     */
    async doUpload(
        filePath,
        name,
        description,
        tier,
        category,
        scope,
        rank,
        alpha,
        onProgress
    ) {
        const stats = fs.statSync(filePath);
        const fileSize = stats.size;

        // Adjust timeout based on file size (1 second per MB minimum)
        const adaptiveTimeout = Math.max(
            this.timeout,
            (fileSize / (1024 * 1024)) * 1000
        );

        const form = new FormData();

        // Create read stream with progress tracking
        const fileStream = fs.createReadStream(filePath);
        let uploadedBytes = 0;

        fileStream.on('data', (chunk) => {
            uploadedBytes += chunk.length;
            if (onProgress) {
                onProgress(uploadedBytes, fileSize);
            }
        });

        form.append('file', fileStream, {
            filename: path.basename(filePath),
        });
        form.append('name', name);
        form.append('tier', tier);
        form.append('category', category);
        form.append('scope', scope);
        form.append('rank', String(rank));
        form.append('alpha', String(alpha));

        if (description) {
            form.append('description', description);
        }

        const controller = new AbortController();
        const timeout = setTimeout(
            () => controller.abort(),
            adaptiveTimeout
        );

        try {
            const response = await fetch(
                `${this.baseUrl}/v1/adapters/upload-aos`,
                {
                    method: 'POST',
                    headers: {
                        'Authorization': `Bearer ${this.token}`,
                        ...form.getHeaders(),
                    },
                    body: form,
                    signal: controller.signal,
                }
            );

            // Handle HTTP errors
            if (response.status === 400) {
                const error = await response.json();
                throw new ValueError(
                    `Validation error (${error.error_code}): ${error.message}`
                );
            } else if (response.status === 403) {
                throw new PermissionError(
                    'Insufficient permissions (need Admin or Operator role)'
                );
            } else if (response.status === 409) {
                throw new Error('Adapter ID conflict (UUID collision, retry)');
            } else if (response.status === 413) {
                throw new Error('File too large for endpoint');
            } else if (response.status === 507) {
                throw new Error('Server disk space exhausted');
            }

            if (!response.ok) {
                const error = await response.json();
                throw new Error(
                    `Upload failed (${response.status}): ` +
                    `${error.error_code} - ${error.message}`
                );
            }

            return response.json();
        } finally {
            clearTimeout(timeout);
        }
    }

    sleep(ms) {
        return new Promise((resolve) => setTimeout(resolve, ms));
    }
}

/**
 * Create minimal valid .aos file for testing
 */
function createTestAosFile(filePath) {
    const manifest = {
        version: '1.0.0',
        model_type: 'lora',
        base_model: 'llama',
        rank: 16,
        alpha: 8.0,
    };

    const manifestJson = JSON.stringify(manifest);
    const manifestBuffer = Buffer.from(manifestJson, 'utf8');

    const buffer = Buffer.alloc(8 + manifestBuffer.length);

    // Write header (little-endian)
    buffer.writeUInt32LE(8, 0); // Manifest offset
    buffer.writeUInt32LE(manifestBuffer.length, 4); // Manifest length

    // Write manifest
    manifestBuffer.copy(buffer, 8);

    fs.writeFileSync(filePath, buffer);
}

/**
 * Example 1: Simple upload
 */
async function exampleSimpleUpload() {
    console.log('='.repeat(60));
    console.log('Example 1: Simple Upload');
    console.log('='.repeat(60));

    const token = process.env.JWT_TOKEN || 'test-token';
    const apiUrl = process.env.API_URL || 'http://localhost:8080';

    const uploader = new SimpleUploader(apiUrl, token);

    // Create test file
    const testFile = '/tmp/test_adapter.aos';
    createTestAosFile(testFile);

    try {
        const result = await uploader.upload(testFile, 'Simple Test Adapter');
        console.log('✓ Upload successful!');
        console.log(`  Adapter ID: ${result.adapter_id}`);
        console.log(`  Hash: ${result.hash_b3}`);
        console.log(`  State: ${result.lifecycle_state}`);
    } catch (error) {
        console.error(`✗ Upload failed: ${error.message}`);
    } finally {
        if (fs.existsSync(testFile)) {
            fs.unlinkSync(testFile);
        }
    }
}

/**
 * Example 2: Production upload with validation
 */
async function exampleProductionUpload() {
    console.log('\n' + '='.repeat(60));
    console.log('Example 2: Production Upload');
    console.log('='.repeat(60));

    const token = process.env.JWT_TOKEN || 'test-token';
    const apiUrl = process.env.API_URL || 'http://localhost:8080';

    const uploader = new ProductionUploader(apiUrl, token, {
        maxRetries: 3,
        timeout: 60000,
    });

    // Create test file
    const testFile = '/tmp/production_adapter.aos';
    createTestAosFile(testFile);

    try {
        const result = await uploader.upload(
            testFile,
            'Production Test Adapter',
            {
                description: 'Tested adapter for production',
                tier: 'persistent',
                category: 'code',
                rank: 16,
                alpha: 8.0,
                onProgress: (done, total) => {
                    const pct = ((done / total) * 100).toFixed(1);
                    process.stdout.write(`  Progress: ${pct}%\r`);
                },
            }
        );

        console.log('\n✓ Upload successful!');
        console.log(`  Adapter ID: ${result.adapter_id}`);
        console.log(`  File size: ${result.file_size} bytes`);
        console.log(`  Hash: ${result.hash_b3}`);
        console.log(`  State: ${result.lifecycle_state}`);
    } catch (error) {
        console.error(`✗ Upload failed: ${error.message}`);
    } finally {
        if (fs.existsSync(testFile)) {
            fs.unlinkSync(testFile);
        }
    }
}

/**
 * Example 3: Batch upload with rate limiting
 */
async function exampleBatchUpload() {
    console.log('\n' + '='.repeat(60));
    console.log('Example 3: Batch Upload');
    console.log('='.repeat(60));

    const token = process.env.JWT_TOKEN || 'test-token';
    const apiUrl = process.env.API_URL || 'http://localhost:8080';

    const uploader = new ProductionUploader(apiUrl, token);

    // Create multiple test files
    const testFiles = [];
    for (let i = 0; i < 3; i++) {
        const testFile = `/tmp/batch_adapter_${i}.aos`;
        createTestAosFile(testFile);
        testFiles.push(testFile);
    }

    try {
        const results = [];
        for (let i = 0; i < testFiles.length; i++) {
            console.log(`Uploading batch_adapter_${i}.aos...`);
            const result = await uploader.upload(
                testFiles[i],
                `Batch Adapter ${i}`,
                { tier: 'ephemeral' }
            );
            results.push(result);
            console.log(`  ✓ ${result.adapter_id}`);

            // Rate limiting: ~2 uploads per second
            if (i < testFiles.length - 1) {
                await new Promise((resolve) => setTimeout(resolve, 500));
            }
        }

        console.log(
            `\n✓ Batch upload complete (${results.length} adapters)`
        );
        results.forEach((result) => {
            console.log(
                `  - ${result.adapter_id}: ${result.lifecycle_state}`
            );
        });
    } catch (error) {
        console.error(`✗ Batch upload failed: ${error.message}`);
    } finally {
        for (const testFile of testFiles) {
            if (fs.existsSync(testFile)) {
                fs.unlinkSync(testFile);
            }
        }
    }
}

/**
 * Example 4: Error handling patterns
 */
async function exampleErrorHandling() {
    console.log('\n' + '='.repeat(60));
    console.log('Example 4: Error Handling');
    console.log('='.repeat(60));

    const token = process.env.JWT_TOKEN || 'test-token';
    const apiUrl = process.env.API_URL || 'http://localhost:8080';

    const uploader = new ProductionUploader(apiUrl, token);

    // Example 4a: File validation errors
    console.log('\n4a. Testing validation errors:');

    try {
        uploader.validateFile('/nonexistent/file.aos');
    } catch (error) {
        console.log(`  ✓ Caught: ${error.message}`);
    }

    try {
        const testFile = '/tmp/wrong_ext.txt';
        fs.writeFileSync(testFile, 'test');
        uploader.validateFile(testFile);
    } catch (error) {
        console.log(`  ✓ Caught: ${error.message}`);
    }

    try {
        await uploader.upload('valid.aos', 'Test', { rank: 1000 });
    } catch (error) {
        console.log(`  ✓ Caught: ${error.message}`);
    }

    // Example 4b: Network errors (conceptual)
    console.log('\n4b. Network retry strategy:');
    console.log('  Timeout → Wait 2s → Retry');
    console.log('  Connection error → Wait 4s → Retry');
    console.log('  After 3 attempts → Fail with descriptive error');
}

/**
 * Main entry point
 */
async function main() {
    console.log('\nAdapterOS Upload Examples');
    console.log(
        'Set JWT_TOKEN and API_URL environment variables\n'
    );

    try {
        await exampleSimpleUpload();
        await exampleProductionUpload();
        await exampleBatchUpload();
        await exampleErrorHandling();

        console.log('\n' + '='.repeat(60));
        console.log('Examples complete!');
        console.log('='.repeat(60));
    } catch (error) {
        console.error('Fatal error:', error);
        process.exit(1);
    }
}

// Run examples if this is the main module
if (require.main === module) {
    main().catch(console.error);
}

module.exports = { SimpleUploader, ProductionUploader, createTestAosFile };

File: examples/upload_examples.py
#!/usr/bin/env python3
"""
AdapterOS Upload Examples

Complete working examples for uploading .aos adapters to AdapterOS.
Run with: python upload_examples.py
"""

import requests
import json
import time
import os
from pathlib import Path
from dataclasses import dataclass
from typing import Optional, Callable
import struct


@dataclass
class UploadResponse:
    """Parsed upload response"""
    adapter_id: str
    tenant_id: str
    hash_b3: str
    file_path: str
    file_size: int
    lifecycle_state: str
    created_at: str


class SimpleUploader:
    """Minimal upload example"""

    def __init__(self, base_url: str, token: str):
        self.base_url = base_url
        self.token = token

    def upload(self, file_path: str, name: str) -> UploadResponse:
        """Upload .aos file"""
        with open(file_path, 'rb') as f:
            response = requests.post(
                f'{self.base_url}/v1/adapters/upload-aos',
                headers={'Authorization': f'Bearer {self.token}'},
                files={
                    'file': f,
                    'name': (None, name),
                },
            )

        response.raise_for_status()
        data = response.json()

        return UploadResponse(
            adapter_id=data['adapter_id'],
            tenant_id=data['tenant_id'],
            hash_b3=data['hash_b3'],
            file_path=data['file_path'],
            file_size=data['file_size'],
            lifecycle_state=data['lifecycle_state'],
            created_at=data['created_at'],
        )


class ProductionUploader:
    """Production-grade uploader with retry, validation, and error handling"""

    def __init__(
        self,
        base_url: str,
        token: str,
        max_retries: int = 3,
        timeout: float = 60.0,
    ):
        self.base_url = base_url
        self.token = token
        self.max_retries = max_retries
        self.timeout = timeout
        self.session = self._create_session()

    def _create_session(self) -> requests.Session:
        """Create optimized session with connection pooling and retry logic"""
        session = requests.Session()

        # Configure connection pooling
        from requests.adapters import HTTPAdapter
        from urllib3.util.retry import Retry

        retry_strategy = Retry(
            total=self.max_retries,
            backoff_factor=1.0,
            status_forcelist=[500, 502, 503, 504],
            allowed_methods=["HEAD", "GET", "OPTIONS", "POST"]
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        session.mount("http://", adapter)
        session.mount("https://", adapter)

        return session

    def validate_file(self, file_path: str) -> None:
        """Validate file before upload"""
        path = Path(file_path)

        # Check existence
        if not path.exists():
            raise FileNotFoundError(f"File not found: {file_path}")

        # Check extension
        if path.suffix != '.aos':
            raise ValueError(f"File must have .aos extension, got {path.suffix}")

        # Check size
        file_size = path.stat().st_size
        max_size = 1024 * 1024 * 1024  # 1GB
        if file_size > max_size:
            raise ValueError(
                f"File too large: {file_size / (1024*1024):.1f}MB "
                f"(max: {max_size / (1024*1024):.0f}MB)"
            )

        # Validate .aos structure
        try:
            self._validate_aos_structure(file_path)
        except Exception as e:
            raise ValueError(f"Invalid .aos file: {e}")

    def _validate_aos_structure(self, file_path: str) -> None:
        """Validate .aos file format"""
        with open(file_path, 'rb') as f:
            # Read header
            header = f.read(8)
            if len(header) < 8:
                raise ValueError("File too small for .aos header")

            offset, length = struct.unpack('<II', header)

            # Validate header values
            if offset < 8 or length == 0:
                raise ValueError(
                    f"Invalid .aos header: offset={offset}, length={length}"
                )

            # Check manifest fits within file
            f.seek(0, 2)  # Seek to end
            file_size = f.tell()

            if offset + length > file_size:
                raise ValueError(
                    f"Manifest extends beyond file: "
                    f"offset={offset}, length={length}, file_size={file_size}"
                )

            # Validate manifest is JSON
            f.seek(offset)
            manifest_bytes = f.read(length)

            try:
                manifest = json.loads(manifest_bytes)
                if not isinstance(manifest, dict):
                    raise ValueError("Manifest must be JSON object")
            except json.JSONDecodeError as e:
                raise ValueError(f"Manifest is not valid JSON: {e}")

    def upload(
        self,
        file_path: str,
        name: str,
        description: Optional[str] = None,
        tier: str = "ephemeral",
        category: str = "general",
        scope: str = "general",
        rank: int = 1,
        alpha: float = 1.0,
        on_progress: Optional[Callable[[int, int], None]] = None,
    ) -> UploadResponse:
        """
        Upload adapter with validation and retry logic

        Args:
            file_path: Path to .aos file
            name: Display name
            description: Optional description
            tier: ephemeral, warm, or persistent
            category: general, code, text, vision, or audio
            scope: general, public, private, or tenant
            rank: LoRA rank (1-512)
            alpha: LoRA scaling (0.0-100.0)
            on_progress: Optional progress callback (bytes_done, total)

        Returns:
            UploadResponse with adapter metadata

        Raises:
            FileNotFoundError: File doesn't exist
            ValueError: Validation failed
            requests.HTTPError: Upload failed
        """
        # Validate input
        self.validate_file(file_path)

        if not 1 <= rank <= 512:
            raise ValueError(f"Rank must be 1-512, got {rank}")

        if not 0.0 <= alpha <= 100.0:
            raise ValueError(f"Alpha must be 0.0-100.0, got {alpha}")

        # Retry loop
        last_error = None
        for attempt in range(1, self.max_retries + 1):
            try:
                return self._do_upload(
                    file_path, name, description, tier, category, scope,
                    rank, alpha, on_progress
                )
            except requests.exceptions.Timeout:
                last_error = f"Timeout on attempt {attempt}"
                if attempt < self.max_retries:
                    delay = 2 ** attempt
                    print(f"Timeout, retrying in {delay}s...")
                    time.sleep(delay)
            except requests.exceptions.ConnectionError:
                last_error = f"Connection error on attempt {attempt}"
                if attempt < self.max_retries:
                    delay = 2 ** attempt
                    print(f"Connection error, retrying in {delay}s...")
                    time.sleep(delay)

        raise RuntimeError(f"Upload failed: {last_error}")

    def _do_upload(
        self,
        file_path: str,
        name: str,
        description: Optional[str],
        tier: str,
        category: str,
        scope: str,
        rank: int,
        alpha: float,
        on_progress: Optional[Callable[[int, int], None]],
    ) -> UploadResponse:
        """Perform actual upload"""
        file_size = Path(file_path).stat().st_size
        self.timeout = max(60, file_size / (1024 * 1024))  # 1s per MB minimum

        with open(file_path, 'rb') as f:
            files = {
                'file': f,
                'name': (None, name),
                'tier': (None, tier),
                'category': (None, category),
                'scope': (None, scope),
                'rank': (None, str(rank)),
                'alpha': (None, str(alpha)),
            }

            if description:
                files['description'] = (None, description)

            # Track progress
            if on_progress:
                bytes_read = 0

                def wrapped_read(size=-1):
                    nonlocal bytes_read
                    chunk = f.read(size)
                    bytes_read += len(chunk)
                    on_progress(bytes_read, file_size)
                    return chunk

                f.read = wrapped_read

            response = self.session.post(
                f'{self.base_url}/v1/adapters/upload-aos',
                headers={'Authorization': f'Bearer {self.token}'},
                files=files,
                timeout=self.timeout,
            )

        # Handle errors
        if response.status_code == 400:
            error = response.json()
            raise ValueError(
                f"Validation error ({error.get('error_code')}): "
                f"{error.get('message')}"
            )
        elif response.status_code == 403:
            raise PermissionError(
                "Insufficient permissions (need Admin or Operator role)"
            )
        elif response.status_code == 409:
            raise ValueError("Adapter ID conflict (UUID collision, retry)")
        elif response.status_code == 413:
            raise ValueError("File too large for endpoint")
        elif response.status_code == 507:
            raise RuntimeError("Server disk space exhausted")

        response.raise_for_status()

        # Parse response
        data = response.json()
        return UploadResponse(
            adapter_id=data['adapter_id'],
            tenant_id=data['tenant_id'],
            hash_b3=data['hash_b3'],
            file_path=data['file_path'],
            file_size=data['file_size'],
            lifecycle_state=data['lifecycle_state'],
            created_at=data['created_at'],
        )


def create_test_aos_file(path: str) -> None:
    """Create minimal valid .aos file for testing"""
    manifest = json.dumps({
        "version": "1.0.0",
        "model_type": "lora",
        "base_model": "llama",
        "rank": 16,
        "alpha": 8.0,
    }).encode('utf-8')

    with open(path, 'wb') as f:
        # Write header
        f.write(struct.pack('<I', 8))  # Manifest offset
        f.write(struct.pack('<I', len(manifest)))  # Manifest length
        # Write manifest
        f.write(manifest)


def example_simple_upload():
    """Example 1: Minimal upload"""
    print("=" * 60)
    print("Example 1: Simple Upload")
    print("=" * 60)

    token = os.getenv('JWT_TOKEN', 'test-token')
    api_url = os.getenv('API_URL', 'http://localhost:8080')

    uploader = SimpleUploader(api_url, token)

    # Create test file
    test_file = '/tmp/test_adapter.aos'
    create_test_aos_file(test_file)

    try:
        result = uploader.upload(test_file, 'Simple Test Adapter')
        print(f"✓ Upload successful!")
        print(f"  Adapter ID: {result.adapter_id}")
        print(f"  Hash: {result.hash_b3}")
        print(f"  State: {result.lifecycle_state}")
    except Exception as e:
        print(f"✗ Upload failed: {e}")
    finally:
        os.remove(test_file)


def example_production_upload():
    """Example 2: Production upload with validation"""
    print("\n" + "=" * 60)
    print("Example 2: Production Upload")
    print("=" * 60)

    token = os.getenv('JWT_TOKEN', 'test-token')
    api_url = os.getenv('API_URL', 'http://localhost:8080')

    uploader = ProductionUploader(api_url, token, max_retries=3)

    # Create test file
    test_file = '/tmp/production_adapter.aos'
    create_test_aos_file(test_file)

    try:
        # Upload with progress tracking
        def show_progress(done, total):
            pct = (done / total) * 100
            print(f"  Progress: {pct:.1f}%", end='\r')

        result = uploader.upload(
            test_file,
            name='Production Test Adapter',
            description='Tested adapter for production',
            tier='persistent',
            category='code',
            rank=16,
            alpha=8.0,
            on_progress=show_progress,
        )

        print(f"\n✓ Upload successful!")
        print(f"  Adapter ID: {result.adapter_id}")
        print(f"  File size: {result.file_size} bytes")
        print(f"  Hash: {result.hash_b3}")
        print(f"  State: {result.lifecycle_state}")

    except ValueError as e:
        print(f"✗ Validation error: {e}")
    except PermissionError as e:
        print(f"✗ Permission error: {e}")
    except Exception as e:
        print(f"✗ Upload failed: {e}")
    finally:
        os.remove(test_file)


def example_batch_upload():
    """Example 3: Batch upload with rate limiting"""
    print("\n" + "=" * 60)
    print("Example 3: Batch Upload")
    print("=" * 60)

    token = os.getenv('JWT_TOKEN', 'test-token')
    api_url = os.getenv('API_URL', 'http://localhost:8080')

    uploader = ProductionUploader(api_url, token)

    # Create multiple test files
    test_files = []
    for i in range(3):
        test_file = f'/tmp/batch_adapter_{i}.aos'
        create_test_aos_file(test_file)
        test_files.append((test_file, f'Batch Adapter {i}'))

    try:
        results = []
        for file_path, name in test_files:
            print(f"Uploading {name}...")
            result = uploader.upload(file_path, name, tier='ephemeral')
            results.append(result)
            print(f"  ✓ {result.adapter_id}")
            time.sleep(0.5)  # Rate limit: ~2 uploads/sec

        print(f"\n✓ Batch upload complete ({len(results)} adapters)")
        for result in results:
            print(f"  - {result.adapter_id}: {result.lifecycle_state}")

    except Exception as e:
        print(f"✗ Batch upload failed: {e}")
    finally:
        for file_path, _ in test_files:
            if os.path.exists(file_path):
                os.remove(file_path)


def example_error_handling():
    """Example 4: Error handling patterns"""
    print("\n" + "=" * 60)
    print("Example 4: Error Handling")
    print("=" * 60)

    token = os.getenv('JWT_TOKEN', 'test-token')
    api_url = os.getenv('API_URL', 'http://localhost:8080')

    uploader = ProductionUploader(api_url, token)

    # Example 4a: File validation errors
    print("\n4a. Testing validation errors:")
    try:
        uploader.upload('/nonexistent/file.aos', 'Test')
    except FileNotFoundError as e:
        print(f"  ✓ Caught: {e}")

    try:
        test_file = '/tmp/wrong_ext.txt'
        open(test_file, 'w').write('test')
        uploader.upload(test_file, 'Test')
    except ValueError as e:
        print(f"  ✓ Caught: {e}")
    finally:
        os.remove(test_file)

    try:
        uploader.upload('valid.aos', 'Test', rank=1000)  # Out of bounds
    except ValueError as e:
        print(f"  ✓ Caught: {e}")

    # Example 4b: Network errors (simulated)
    print("\n4b. Network retry strategy:")
    print("  Timeout → Wait 2s → Retry")
    print("  Connection error → Wait 4s → Retry")
    print("  After 3 attempts → Fail with descriptive error")


if __name__ == '__main__':
    print("\nAdapterOS Upload Examples")
    print("Set JWT_TOKEN and API_URL environment variables\n")

    example_simple_upload()
    example_production_upload()
    example_batch_upload()
    example_error_handling()

    print("\n" + "=" * 60)
    print("Examples complete!")
    print("=" * 60)
